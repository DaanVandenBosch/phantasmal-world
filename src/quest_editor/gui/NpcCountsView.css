.quest_editor_NpcCountsView {
    user-select: text;
    box-sizing: border-box;
    padding: 3px;
    overflow: auto;
}

.quest_editor_NpcCountsView table {
    width: 100%;
}

.quest_editor_NpcCountsView th {
    cursor: text;
    text-align: left;
}

.quest_editor_NpcCountsView td {
    cursor: text;
}

.quest_editor_NpcCountsView_no_quest {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
}
