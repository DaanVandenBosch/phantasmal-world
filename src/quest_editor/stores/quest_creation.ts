import { Episode } from "../../core/data_formats/parsing/quest/Episode";
import { QuestModel } from "../model/QuestModel";
import { ObjectType } from "../../core/data_formats/parsing/quest/object_types";
import { NpcType } from "../../core/data_formats/parsing/quest/npc_types";
import { QuestObjectModel } from "../model/QuestObjectModel";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { Euler, Vector3 } from "three";
import { QuestEventDagModel } from "../model/QuestEventDagModel";
import { AreaStore } from "./AreaStore";
import { assemble } from "../scripting/assembly";
import { Segment } from "../scripting/instructions";

export function create_new_quest(area_store: AreaStore, episode: Episode): QuestModel {
    if (episode === Episode.II) throw new Error("Episode II not yet supported.");
    if (episode === Episode.IV) throw new Error("Episode IV not yet supported.");

    return new QuestModel(
        area_store,
        0,
        0,
        "Untitled",
        "Created with phantasmal.world.",
        "Created with phantasmal.world.",
        episode,
        new Map().set(0, 0),
        create_default_objects(),
        create_default_npcs(),
        create_default_event_chains(),
        [],
        create_default_object_code(),
        [],
    );
}

function create_default_objects(): QuestObjectModel[] {
    return [
        new QuestObjectModel(
            ObjectType.MenuActivation,
            16384,
            0,
            0,
            10,
            new Vector3(-16.313568115234375, 3, -579.5118408203125),
            new Euler(0.0009587526218325454, 0, 0, "ZXY"),
            new Map([
                ["property_0", 1],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365279104],
            ]),
            [
                [2, 0, 0, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.MenuActivation,
            16385,
            0,
            0,
            10,
            new Vector3(-393.07318115234375, 10, -12.964752197265625),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 9.183549615799121e-41],
                ["property_1", 1.0000011920928955],
                ["property_2", 1],
                ["property_3", 0],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365279264],
            ]),
            [
                [2, 0, 1, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.MenuActivation,
            16386,
            0,
            0,
            10,
            new Vector3(-458.60699462890625, 10, -51.270660400390625),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 1],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 2],
                ["property_4", 65536],
                ["property_5", 10],
                ["property_6", 2365279424],
            ]),
            [
                [2, 0, 2, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.MenuActivation,
            16387,
            0,
            0,
            10,
            new Vector3(-430.19696044921875, 10, -24.490447998046875),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 1],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 3],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365279584],
            ]),
            [
                [2, 0, 3, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PlayerSet,
            16388,
            0,
            0,
            10,
            new Vector3(0.995330810546875, 0, -37.0010986328125),
            new Euler(0, 4.712460886831327, 0, "ZXY"),
            new Map([
                ["property_0", 0],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 0],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365279744],
            ]),
            [
                [2, 0, 4, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PlayerSet,
            16389,
            0,
            0,
            10,
            new Vector3(3.0009307861328125, 0, -23.99688720703125),
            new Euler(0, 4.859725289544806, 0, "ZXY"),
            new Map([
                ["property_0", 1.000000238418579],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 0],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365279904],
            ]),
            [
                [2, 0, 5, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PlayerSet,
            16390,
            0,
            0,
            10,
            new Vector3(2.0015106201171875, 0, -50.00386047363281),
            new Euler(0, 4.565196484117848, 0, "ZXY"),
            new Map([
                ["property_0", 2.000002384185791],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 0],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365280064],
            ]),
            [
                [2, 0, 6, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PlayerSet,
            16391,
            0,
            0,
            10,
            new Vector3(4.9973907470703125, 0, -61.99664306640625),
            new Euler(0, 4.368843947166543, 0, "ZXY"),
            new Map([
                ["property_0", 3.0000007152557373],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 65536],
                ["property_4", 10],
                ["property_5", 0],
                ["property_6", 2365280224],
            ]),
            [
                [2, 0, 7, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.MainRagolTeleporter,
            18264,
            0,
            0,
            10,
            new Vector3(132.00314331054688, 1.000000238418579, -265.002197265625),
            new Euler(0, 0.49088134237826325, 0, "ZXY"),
            new Map([
                ["property_0", 1.000000238418579],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365227216],
            ]),
            [
                [0, 0, 87, 7, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PrincipalWarp,
            16393,
            0,
            0,
            10,
            new Vector3(-228, 0, -2020.99951171875),
            new Euler(0, 2.9452880542695796, 0, "ZXY"),
            new Map([
                ["property_0", -10.000004768371582],
                ["property_1", 0],
                ["property_2", -30.000030517578125],
                ["property_3", 0],
                ["property_4", 65536],
                ["property_5", 65536],
                ["property_6", 2365280688],
            ]),
            [
                [2, 0, 9, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.MenuActivation,
            16394,
            0,
            0,
            10,
            new Vector3(-41.000030517578125, 0, 42.37322998046875),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 1],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 4],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365280992],
            ]),
            [
                [2, 0, 10, 0, 0, 0],
                [1, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.MenuActivation,
            16395,
            0,
            0,
            10,
            new Vector3(-479.21673583984375, 8.781256675720215, -322.465576171875),
            new Euler(6.28328118244177, 0.0009587526218325454, 0, "ZXY"),
            new Map([
                ["property_0", 1],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 5],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365281152],
            ]),
            [
                [2, 0, 11, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PrincipalWarp,
            16396,
            0,
            0,
            10,
            new Vector3(-228, 0, -351.0015869140625),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 10.000006675720215],
                ["property_1", 0],
                ["property_2", -1760.0010986328125],
                ["property_3", 32768],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365281312],
            ]),
            [
                [2, 0, 12, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.TelepipeLocation,
            16397,
            0,
            0,
            10,
            new Vector3(-561.88232421875, 0, -406.8829345703125),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 1],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 0],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365281616],
            ]),
            [
                [2, 0, 13, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.TelepipeLocation,
            16398,
            0,
            0,
            10,
            new Vector3(-547.8557739257812, 0, -444.8822326660156),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 1],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365281808],
            ]),
            [
                [2, 0, 14, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.TelepipeLocation,
            16399,
            0,
            0,
            10,
            new Vector3(-486.441650390625, 0, -497.4501647949219),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 9.183549615799121e-41],
                ["property_1", 1.0000011920928955],
                ["property_2", 1],
                ["property_3", 3],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365282000],
            ]),
            [
                [2, 0, 15, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.TelepipeLocation,
            16400,
            0,
            0,
            10,
            new Vector3(-522.4052734375, 0, -474.1882629394531),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 1],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 2],
                ["property_4", 65536],
                ["property_5", 10],
                ["property_6", 2365282192],
            ]),
            [
                [2, 0, 16, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.MedicalCenterDoor,
            16401,
            0,
            0,
            10,
            new Vector3(-34.49853515625, 0, -384.4951171875),
            new Euler(0, 5.497871034636549, 0, "ZXY"),
            new Map([
                ["property_0", 3.0000007152557373],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365282384],
            ]),
            [
                [2, 0, 17, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.ShopDoor,
            16402,
            0,
            0,
            10,
            new Vector3(-393.0031433105469, 0, -143.49981689453125),
            new Euler(0, 3.141640591220885, 0, "ZXY"),
            new Map([
                ["property_0", 3.0000007152557373],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365282640],
            ]),
            [
                [2, 0, 18, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.MenuActivation,
            16403,
            0,
            0,
            10,
            new Vector3(-355.17462158203125, 0, -43.15193176269531),
            new Euler(0, 0, 0, "ZXY"),
            new Map([
                ["property_0", 1.000000238418579],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 6],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365282896],
            ]),
            [
                [2, 0, 19, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.HuntersGuildDoor,
            16404,
            0,
            0,
            10,
            new Vector3(-43.00239562988281, 0, -118.00120544433594),
            new Euler(0, 3.141640591220885, 0, "ZXY"),
            new Map([
                ["property_0", 3.0000007152557373],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365283056],
            ]),
            [
                [2, 0, 20, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.TeleporterDoor,
            16405,
            0,
            0,
            10,
            new Vector3(26.000823974609375, 0, -265.99810791015625),
            new Euler(0, 3.141640591220885, 0, "ZXY"),
            new Map([
                ["property_0", 3.0000007152557373],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365283312],
            ]),
            [
                [2, 0, 21, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PlayerSet,
            16406,
            0,
            0,
            10,
            new Vector3(57.81005859375, 0, -268.5472412109375),
            new Euler(0, 4.712460886831327, 0, "ZXY"),
            new Map([
                ["property_0", 0],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365283568],
            ]),
            [
                [2, 0, 22, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PlayerSet,
            16407,
            0,
            0,
            10,
            new Vector3(66.769287109375, 0, -252.3748779296875),
            new Euler(0, 4.712460886831327, 0, "ZXY"),
            new Map([
                ["property_0", 1.000000238418579],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365283728],
            ]),
            [
                [2, 0, 23, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PlayerSet,
            16408,
            0,
            0,
            10,
            new Vector3(67.36819458007812, 0, -284.9297180175781),
            new Euler(0, 4.712460886831327, 0, "ZXY"),
            new Map([
                ["property_0", 2.000000476837158],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365283888],
            ]),
            [
                [2, 0, 24, 0, 0, 0],
                [0, 0],
            ],
        ),
        new QuestObjectModel(
            ObjectType.PlayerSet,
            16409,
            0,
            0,
            10,
            new Vector3(77.10488891601562, 0, -269.2830505371094),
            new Euler(0, 4.712460886831327, 0, "ZXY"),
            new Map([
                ["property_0", 3.0000007152557373],
                ["property_1", 1],
                ["property_2", 1],
                ["property_3", 1],
                ["property_4", 0],
                ["property_5", 0],
                ["property_6", 2365284048],
            ]),
            [
                [2, 0, 25, 0, 0, 0],
                [0, 0],
            ],
        ),
    ];
}

function create_default_npcs(): QuestNpcModel[] {
    return [
        new QuestNpcModel(
            NpcType.GuildLady,
            29,
            1011.0010986328125,
            0,
            0,
            850,
            0,
            0,
            10,
            new Vector3(-49.0010986328125, 0, 50.996429443359375),
            new Euler(0, 2.3562304434156633, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 86, 0, 0, 0, 0, 23, 87],
                [128, 238, 223, 176],
            ],
        ),
        new QuestNpcModel(
            NpcType.FemaleFat,
            4,
            1016.0010986328125,
            0,
            0,
            100,
            1,
            0,
            20,
            new Vector3(167.99769592285156, 0, 83.99686431884766),
            new Euler(0, 3.927050739026106, 0, "ZXY"),
            new Vector3(24.000009536743164, 0, 0),
            [
                [0, 0, 7, 88, 0, 0, 0, 0, 23, 89],
                [128, 238, 232, 48],
            ],
        ),
        new QuestNpcModel(
            NpcType.MaleDwarf,
            10,
            1015.0010986328125,
            0,
            0,
            90,
            1,
            0,
            20,
            new Vector3(156.0028839111328, 0, -49.99967575073242),
            new Euler(0, 5.497871034636549, 0, "ZXY"),
            new Vector3(30.000009536743164, 0, 0),
            [
                [0, 0, 7, 89, 0, 0, 0, 0, 23, 90],
                [128, 238, 236, 176],
            ],
        ),
        new QuestNpcModel(
            NpcType.RedSoldier,
            26,
            1020.0010986328125,
            0,
            0,
            130,
            0,
            0,
            20,
            new Vector3(237.9988250732422, 0, -14.0001220703125),
            new Euler(0, 5.497871034636549, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 90, 0, 0, 0, 0, 23, 91],
                [128, 238, 241, 48],
            ],
        ),
        new QuestNpcModel(
            NpcType.BlueSoldier,
            25,
            1019.0010986328125,
            0,
            0,
            120,
            0,
            0,
            20,
            new Vector3(238.00379943847656, 0, 63.00413513183594),
            new Euler(0, 3.927050739026106, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 91, 0, 0, 0, 0, 23, 92],
                [128, 238, 245, 176],
            ],
        ),
        new QuestNpcModel(
            NpcType.FemaleMacho,
            5,
            1014.0010986328125,
            0,
            0,
            80,
            1,
            0,
            20,
            new Vector3(-2.001882553100586, 0, 35.0036506652832),
            new Euler(0, 3.141640591220885, 0, "ZXY"),
            new Vector3(26.000009536743164, 0, 0),
            [
                [0, 0, 7, 92, 0, 0, 0, 0, 23, 93],
                [128, 238, 250, 48],
            ],
        ),
        new QuestNpcModel(
            NpcType.Scientist,
            30,
            1013.0010986328125,
            0,
            0,
            70,
            1,
            0,
            20,
            new Vector3(-147.0000457763672, 0, -7.996537208557129),
            new Euler(0, 2.577127047485882, 0, "ZXY"),
            new Vector3(30.000009536743164, 0, 0),
            [
                [0, 0, 7, 93, 0, 0, 0, 0, 23, 94],
                [128, 238, 254, 176],
            ],
        ),
        new QuestNpcModel(
            NpcType.MaleOld,
            13,
            1012.0010986328125,
            0,
            0,
            60,
            1,
            0,
            20,
            new Vector3(-219.99710083007812, 0, -100.0008316040039),
            new Euler(0, 0, 0, "ZXY"),
            new Vector3(30.000011444091797, 0, 0),
            [
                [0, 0, 7, 94, 0, 0, 0, 0, 23, 95],
                [128, 239, 3, 48],
            ],
        ),
        new QuestNpcModel(
            NpcType.GuildLady,
            29,
            1010.0010986328125,
            0,
            0,
            840,
            0,
            0,
            20,
            new Vector3(-262.5099792480469, 0, -24.53999900817871),
            new Euler(0, 1.963525369513053, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 95, 0, 0, 0, 0, 23, 106],
                [128, 239, 100, 192],
            ],
        ),
        new QuestNpcModel(
            NpcType.Tekker,
            28,
            1009,
            0,
            0,
            830,
            0,
            0,
            30,
            new Vector3(-43.70983123779297, 2.5999999046325684, -52.78248596191406),
            new Euler(0, 0.7854101478052212, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 97, 0, 0, 0, 0, 23, 98],
                [128, 239, 16, 176],
            ],
        ),
        new QuestNpcModel(
            NpcType.MaleMacho,
            12,
            1006,
            0,
            0,
            800,
            0,
            0,
            30,
            new Vector3(0.33990478515625, 2.5999999046325684, -84.71995544433594),
            new Euler(0, 0, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 98, 0, 0, 0, 0, 23, 99],
                [128, 239, 21, 48],
            ],
        ),
        new QuestNpcModel(
            NpcType.FemaleMacho,
            5,
            1008,
            0,
            0,
            820,
            0,
            0,
            30,
            new Vector3(43.87113952636719, 2.5999996662139893, -74.80299377441406),
            new Euler(0, -0.5645135437350027, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 99, 0, 0, 0, 0, 23, 100],
                [128, 239, 25, 176],
            ],
        ),
        new QuestNpcModel(
            NpcType.MaleFat,
            11,
            1007.0010986328125,
            0,
            0,
            810,
            0,
            0,
            30,
            new Vector3(75.88380432128906, 2.5999996662139893, -42.69328308105469),
            new Euler(0, -1.0308508189943528, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 100, 0, 0, 0, 0, 23, 101],
                [128, 239, 30, 48],
            ],
        ),
        new QuestNpcModel(
            NpcType.FemaleTall,
            7,
            1021.0010986328125,
            0,
            0,
            140,
            1,
            0,
            30,
            new Vector3(16.003997802734375, 0, 5.995697021484375),
            new Euler(0, -1.1781152217078317, 0, "ZXY"),
            new Vector3(22.000009536743164, 0, 0),
            [
                [0, 0, 7, 101, 0, 0, 0, 0, 23, 102],
                [128, 239, 34, 176],
            ],
        ),
        new QuestNpcModel(
            NpcType.Nurse,
            31,
            1017,
            0,
            0,
            860,
            0,
            0,
            40,
            new Vector3(0.3097381591796875, 3, -105.3865966796875),
            new Euler(0, 0, 0, "ZXY"),
            new Vector3(0, 0, 0),
            [
                [0, 0, 7, 102, 0, 0, 0, 0, 23, 103],
                [128, 239, 39, 48],
            ],
        ),
        new QuestNpcModel(
            NpcType.Nurse,
            31,
            1018.0010986328125,
            0,
            0,
            110,
            1,
            0,
            40,
            new Vector3(53.499176025390625, 0, -26.496688842773438),
            new Euler(0, 5.497871034636549, 0, "ZXY"),
            new Vector3(18.000009536743164, 0, 0),
            [
                [0, 0, 7, 103, 0, 0, 0, 0, 23, 104],
                [128, 239, 43, 176],
            ],
        ),
    ];
}

function create_default_event_chains(): QuestEventDagModel[] {
    return [];
}

function create_default_object_code(): Segment[] {
    return assemble(
        `.code

0:
    set_episode 0
    set_floor_handler 0, 150
    bb_map_designate 0, 0, 0, 0
    ret
150:
    leti r60, 237
    leti r61, 0
    leti r62, 333
    leti r63, -15
    p_setpos 0, r60
    leti r60, 255
    leti r61, 0
    leti r62, 338
    leti r63, -43
    p_setpos 1, r60
    leti r60, 222
    leti r61, 0
    leti r62, 322
    leti r63, 25
    p_setpos 2, r60
    leti r60, 248
    leti r61, 0
    leti r62, 323
    leti r63, -20
    p_setpos 3, r60
    ret`.split("\n"),
    ).object_code;
}
