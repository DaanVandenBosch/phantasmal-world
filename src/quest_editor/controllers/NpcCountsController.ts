import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { npc_data, NpcType } from "../../core/data_formats/parsing/quest/npc_types";
import { property } from "../../core/observable";

export type NameWithCount = { readonly name: string; readonly count: number };

export class NpcCountsController extends Controller {
    readonly npc_counts: Property<readonly NameWithCount[]>;
    readonly unavailable: Property<boolean>;

    constructor(store: QuestEditorStore) {
        super();

        this.unavailable = store.current_quest.map(q => q == undefined);

        this.npc_counts = store.current_quest
            .flat_map(quest => (quest ? quest.npcs : property([])))
            .map(this.update_view);
    }

    private update_view(npcs: readonly QuestNpcModel[]): NameWithCount[] {
        const npc_counts = new Map<NpcType, number>();

        for (const npc of npcs) {
            const val = npc_counts.get(npc.type) || 0;
            npc_counts.set(npc.type, val + 1);
        }

        const extra_canadines = (npc_counts.get(NpcType.Canane) || 0) * 8;

        // Sort by canonical order.
        const sorted_npc_counts = [...npc_counts].sort((a, b) => a[0] - b[0]);

        return sorted_npc_counts.map(([npc_type, count]) => {
            const extra = npc_type === NpcType.Canadine ? extra_canadines : 0;

            return { name: npc_data(npc_type).name, count: count + extra };
        });
    }
}
