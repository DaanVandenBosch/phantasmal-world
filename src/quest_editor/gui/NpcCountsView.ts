import { ResizableWidget } from "../../core/gui/ResizableWidget";
import { bind_attr, el } from "../../core/gui/dom";
import "./NpcCountsView.css";
import { UnavailableView } from "./UnavailableView";
import { NameWithCount, NpcCountsController } from "../controllers/NpcCountsController";

export class NpcCountsView extends ResizableWidget {
    readonly element = el.div({ class: "quest_editor_NpcCountsView" });

    private readonly table_element = el.table();

    private readonly unavailable_view = new UnavailableView("No quest loaded.");

    constructor(ctrl: NpcCountsController) {
        super();

        this.element.append(this.table_element, this.unavailable_view.element);

        this.disposables(
            bind_attr(this.table_element, "hidden", ctrl.unavailable),

            this.unavailable_view.visible.bind_to(ctrl.unavailable),

            ctrl.npc_counts.observe(({ value }) => this.update_view(value), { call_now: true }),
        );

        this.finalize_construction();
    }

    private update_view(npcs: readonly NameWithCount[]): void {
        const frag = document.createDocumentFragment();

        for (const { name, count } of npcs) {
            frag.append(el.tr({}, el.th({ text: name + ":" }), el.td({ text: String(count) })));
        }

        this.table_element.innerHTML = "";
        this.table_element.append(frag);
    }
}
