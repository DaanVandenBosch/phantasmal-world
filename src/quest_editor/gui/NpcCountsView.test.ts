import { NpcCountsView } from "./NpcCountsView";
import { NpcCountsController } from "../controllers/NpcCountsController";
import {
    create_area_store,
    create_quest_editor_store,
} from "../../../test/src/quest_editor/stores/store_creation";
import { with_disposer } from "../../../test/src/core/observables/disposable_helpers";
import { create_new_quest } from "../stores/quest_creation";
import { Episode } from "../../core/data_formats/parsing/quest/Episode";

test("Renders correctly without a current quest.", () =>
    with_disposer(disposer => {
        const store = create_quest_editor_store(disposer);
        const view = disposer.add(new NpcCountsView(disposer.add(new NpcCountsController(store))));

        expect(view.element).toMatchSnapshot('Should render a "No quest loaded." view.');
    }));

test("Renders correctly with a current quest.", () =>
    with_disposer(disposer => {
        const area_store = create_area_store(disposer);
        const store = create_quest_editor_store(disposer, area_store);
        const view = disposer.add(new NpcCountsView(disposer.add(new NpcCountsController(store))));

        store.set_current_quest(create_new_quest(area_store, Episode.I));

        expect(view.element).toMatchSnapshot("Should render a table with NPC names and counts.");
    }));
