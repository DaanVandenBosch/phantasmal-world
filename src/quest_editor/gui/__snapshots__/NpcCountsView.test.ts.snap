// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with a current quest.: Should render a table with NPC names and counts. 1`] = `
<div
  class="quest_editor_NpcCountsView"
>
  <table>
    <tr>
      <th>
        Female Fat:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Female Macho:
      </th>
      <td>
        2
      </td>
    </tr>
    <tr>
      <th>
        Female Tall:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Male Dwarf:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Male Fat:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Male Macho:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Male Old:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Blue Soldier:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Red Soldier:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Tekker:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Guild Lady:
      </th>
      <td>
        2
      </td>
    </tr>
    <tr>
      <th>
        Scientist:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Nurse:
      </th>
      <td>
        2
      </td>
    </tr>
  </table>
  <div
    class="quest_editor_UnavailableView"
    hidden=""
  >
    <label
      class="core_Label disabled"
    >
      No quest loaded.
    </label>
  </div>
</div>
`;

exports[`Renders correctly with a current quest.: should render a table with NPC names and counts 1`] = `
<div
  class="quest_editor_NpcCountsView"
>
  <table>
    <tr>
      <th>
        Female Fat:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Female Macho:
      </th>
      <td>
        2
      </td>
    </tr>
    <tr>
      <th>
        Female Tall:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Male Dwarf:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Male Fat:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Male Macho:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Male Old:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Blue Soldier:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Red Soldier:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Tekker:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Guild Lady:
      </th>
      <td>
        2
      </td>
    </tr>
    <tr>
      <th>
        Scientist:
      </th>
      <td>
        1
      </td>
    </tr>
    <tr>
      <th>
        Nurse:
      </th>
      <td>
        2
      </td>
    </tr>
  </table>
  <div
    class="quest_editor_UnavailableView"
    hidden=""
  >
    <label
      class="core_Label disabled"
    >
      No quest loaded.
    </label>
  </div>
</div>
`;

exports[`Renders correctly without a current quest.: Should render a "No quest loaded." view. 1`] = `
<div
  class="quest_editor_NpcCountsView"
>
  <table
    hidden=""
  />
  <div
    class="quest_editor_UnavailableView"
  >
    <label
      class="core_Label disabled"
    >
      No quest loaded.
    </label>
  </div>
</div>
`;

exports[`Renders correctly without a current quest.: should render a "No quest loaded." view 1`] = `
<div
  class="quest_editor_NpcCountsView"
>
  <table
    hidden=""
  />
  <div
    class="quest_editor_UnavailableView"
  >
    <label
      class="core_Label disabled"
    >
      No quest loaded.
    </label>
  </div>
</div>
`;
